### PR TITLE
Fix callout json plugin show global stats

### PR DESCRIPTION
**Pull request by daBONDi**
##### SUMMARY

- Add support that callback plugin **json** show also non-host custom stats like you can
define in module **set_stats** with module parameter **per_host: no**

- Add "custom_stats_global" dict  to contain the set_stats with param "per_host:no"

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
callback plugin: json

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (fix_json_callback_plugin_custom_stats 3262f0a15e) last updated 2018/07/18 21:12:11 (GMT +200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/david/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /mnt/g/dev/ansible/lib/ansible
  executable location = /mnt/g/dev/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Fix my need for the non host related custom stat describe in issue/feature #42871

before
```json
    "custom_stats": {
        "localhost": {
            "MyHostVar": "MyHostVarValue"
        },
        "ubuntudev": {
            "MyHostVar": "MyHostVarValue"
        }
    },...
```
after
```json
{
   "custom_stats_global":{
         "MyPlayVar":"MyPlayVarValueMyPlayVarValue",
         "subvalue":{
            "and_a_list":[
               "value1",
               "value2"
            ],
            "sum_more_sub_value":"hihi"
         }
      },
  }
  "custom_stats": {
      "localhost":{
         "MyHostVar":"MyHostVarValue",
      },
      "ubuntudev":{
         "MyHostVar":"MyHostVarValue"
      }
   },
   ...
```

### CODE REVIEW

**Comment by daBONDi** [comment]
Rebased

File: lib/ansible/plugins/callback/json.py [hunk 1]
```diff
@@ -100,15 +100,24 @@ def v2_playbook_on_stats(self, stats):
...
+                    custom_stats['global'] = stats.custom[k]
+                    continue
+
+                custom_stats['hosts'].update({self._convert_host_to_name(k): stats.custom[k]})
 
         output = {
             'plays': self.results,
             'stats': summary,
-            'custom_stats': custom_stats,
+            'custom_stats': custom_stats
```
Comment (sivel) [inline lib/ansible/plugins/callback/json.py]: Please do not remove this trailing comma, this is our preference with dicts and lists to minimize future diffs when lines are changed.

File: lib/ansible/plugins/callback/json.py [hunk 2]
```diff
@@ -100,15 +100,24 @@ def v2_playbook_on_stats(self, stats):
             s = stats.summarize(h)
             summary[h] = s
 
-        custom_stats = {}
+        custom_stats = {
+            'global': {},
+            'hosts': {}
+        }
```
Comment (sivel) [inline lib/ansible/plugins/callback/json.py]: This name spacing change is not backwards compatible and will cause existing uses of this plugin to fail unexpectedly.  I believe a different means will be required here.

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/callback/json.py (hunk 1 above)
Comment (daBONDi) [reply lib/ansible/plugins/callback/json.py]: added it again

File: lib/ansible/plugins/callback/json.py (hunk 2 above)
Comment (daBONDi) [reply lib/ansible/plugins/callback/json.py]: refactor it that custom_stats have the same output as before and add a additional dict custom_stats_global

**Comment by daBONDi** [comment]
I fixed the stuff stivel want me to fix, see new comit.
pls remove need_revision

**Comment by daBONDi** [comment]
Any news to get this into next release ?

File: lib/ansible/plugins/callback/json.py [hunk 3]
```diff
@@ -101,14 +101,22 @@ def v2_playbook_on_stats(self, stats):
             summary[h] = s
 
         custom_stats = {}
+        custom_stats_global = {}
+
         if self.get_option('show_custom_stats') and stats.custom:
-            custom_stats.update(dict((self._convert_host_to_name(k), v) for k, v in stats.custom.items()))
-            custom_stats.pop('_run', None)
+
+            for k in sorted(stats.custom.keys()):
```
Comment (agaffney) [inline lib/ansible/plugins/callback/json.py]: Is the `sorted()` actually useful here? You're adding the keys to a standard python dict, which is unsorted by nature.

Comment (daBONDi) [reply lib/ansible/plugins/callback/json.py]: This was pre pr code i think, so the output of the callback plugin give a ansible_host sorted list.

should i remove it?

Comment (daBONDi) [reply lib/ansible/plugins/callback/json.py]: thanks for the review, i removed it you where right.

File: lib/ansible/plugins/callback/json.py [hunk 4]
```diff
@@ -101,14 +101,22 @@ def v2_playbook_on_stats(self, stats):
...
         if self.get_option('show_custom_stats') and stats.custom:
-            custom_stats.update(dict((self._convert_host_to_name(k), v) for k, v in stats.custom.items()))
-            custom_stats.pop('_run', None)
+
+            for k in stats.custom.keys():
+                if k == '_run':
+                    custom_stats_global = stats.custom[k]
+                    continue
+
+                custom_stats.update({self._convert_host_to_name(k): stats.custom[k]})
```
Comment (sivel) [inline lib/ansible/plugins/callback/json.py]: I feel like this changes too much, for what could be achieved with less:

```diff
diff --git a/lib/ansible/plugins/callback/json.py b/lib/ansible/plugins/callback/json.py
index 3961a78aab..fd344a0cf0 100644
--- a/lib/ansible/plugins/callback/json.py
+++ b/lib/ansible/plugins/callback/json.py
@@ -101,14 +101,16 @@ class CallbackModule(CallbackBase):
             summary[h] = s
 
         custom_stats = {}
+        global_custom_stats = {}
         if self.get_option('show_custom_stats') and stats.custom:
             custom_stats.update(dict((self._convert_host_to_name(k), v) for k, v in stats.custom.items()))
-            custom_stats.pop('_run', None)
+            global_custom_stats.update(custom_stats.pop('_run', {}))
 
         output = {
             'plays': self.results,
             'stats': summary,
             'custom_stats': custom_stats,
+            'global_custom_stats': global_custom_stats,
         }
 
         self._display.display(json.dumps(output, indent=4, sort_keys=True))
```

**Review by sivel — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/callback/json.py (hunk 4 above)
Comment (daBONDi) [reply lib/ansible/plugins/callback/json.py]: Your code is realy more lean and clean, thanks for the review, i changed it to your suggestion

**Review by sivel — CHANGES_REQUESTED** [review]
One additional thing I need is a changelog fragment added to `changelog/fragments` with a `minor_changes:` heading.

**Comment by daBONDi** [comment]
Added changelog fragment, hopefully i do it right :-)

Thanks for guiding me through it

**Comment by daBONDi** [comment]
damn shippable test failed on downloading epel repo info...

how i can rerun it, without making a new comit?

**Review by sivel — APPROVED** [review]

**Pull request merged**